File: iosApp/iosApp/Modules/QRCodeScan/QRCodeScanPresenter.swift
// Created by web3d4v on 21/06/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3lib

enum QRCodeScanPresenterEvent {
    case qrCode(String)
    case dismiss
}

protocol QRCodeScanPresenter {
    func present()
    func handle(_ event: QRCodeScanPresenterEvent)
}

final class DefaultQRCodeScanPresenter {

    private weak var view: QRCodeScanView?
    private let wireframe: QRCodeScanWireframe
    private let interactor: QRCodeScanInteractor
    private let context: QRCodeScanWireframeContext
    
    private var qrCodeDetected = false
    
    init(
        view: QRCodeScanView,
        wireframe: QRCodeScanWireframe,
        interactor: QRCodeScanInteractor,
        context: QRCodeScanWireframeContext
    ) {
        self.view = view
        self.wireframe = wireframe
        self.interactor = interactor
        self.context = context
    }
}

extension DefaultQRCodeScanPresenter: QRCodeScanPresenter {

    func present() {
        updateView(with: nil)
    }

    func handle(_ event: QRCodeScanPresenterEvent) {
        switch event {
        case let .qrCode(qrCode):
            handleQRCode(qrCode)
        case .dismiss:
            wireframe.dismiss()
        }
    }
}

private extension DefaultQRCodeScanPresenter {
    
    func handleQRCode(_ qrCode: String) {
        switch context.type {
        case .`default`:
            // NOTE: This guard is to prevent calling multiple times wireframe.navigate(to: .qrCode())
            // since the view keeps firing multiple times the same code being scanned
            guard !qrCodeDetected else { return }
            qrCodeDetected = true
            wireframe.navigate(to: .qrCode(qrCode))
        case let .network(network):
            guard let address = interactor.validateAddress(
                address: qrCode,
                for: network
            ) else {
                let errorMessage = Localized("qrCodeScan.error.invalid.address", arg: network.name)
                updateView(with: "\(qrCode)\n\(errorMessage)")
                return
            }
            // NOTE: This guard is to prevent calling multiple times wireframe.navigate(to: .qrCode())
            // since the view keeps firing multiple times the same code being scanned
            guard !qrCodeDetected else { return }
            qrCodeDetected = true
            wireframe.navigate(to: .qrCode(address))
        }
    }
}

private extension DefaultQRCodeScanPresenter {
    
    func updateView(with failure: String?) {
        let viewModel = QRCodeScanViewModel(
            title: Localized("qrCodeScan.title"),
            failure: failure
        )
        view?.update(with: viewModel)
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/qrCodeScan/QRCodeScanPresenter.kt
package com.sonsofcrypto.web3walletcore.modules.qrCodeScan

import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3walletcore.extensions.Localized
import com.sonsofcrypto.web3walletcore.modules.qrCodeScan.QRCodeScanWireframeDestination.Dismiss

sealed class QRCodeScanPresenterEvent {
    data class QRCode(val input: String): QRCodeScanPresenterEvent()
    object Dismiss: QRCodeScanPresenterEvent()
}

interface QRCodeScanPresenter {
    fun present()
    fun viewWillAppear()
    fun handle(event: QRCodeScanPresenterEvent)
}

class DefaultQRCodeScanPresenter(
    private val view: WeakRef<QRCodeScanView>,
    private val wireframe: QRCodeScanWireframe,
    private val interactor: QRCodeScanInteractor,
    private val context: QRCodeScanWireframeContext,
): QRCodeScanPresenter {
    private var qrCodeDetected = false

    override fun present() {
        updateView()
    }

    override fun viewWillAppear() {
        qrCodeDetected = false
    }

    override fun handle(event: QRCodeScanPresenterEvent) {
        when (event) {
            is QRCodeScanPresenterEvent.QRCode -> handleQRCodeEvent(event.input)
            is QRCodeScanPresenterEvent.Dismiss -> wireframe.navigate(Dismiss)
        }
    }

    private fun updateView(failure: String? = null) {
        view.get()?.update(viewModel(failure))
    }

    private fun viewModel(failure: String?): QRCodeScanViewModel =
        QRCodeScanViewModel(Localized("qrCodeScan.title"), failure)

    private fun handleQRCodeEvent(input: String) {
        when (val c = context.type) {
            is QRCodeScanWireframeContext.Type.Default -> handleQRCode(input)
            is QRCodeScanWireframeContext.Type.Network -> {
                val address = interactor.validateAddress(input, c.network)
                if (address != null) { handleQRCode(address) }
                else { updateView(Localized("qrCodeScan.error.invalid.address", c.network.name))}
            }
        }
    }

    private fun handleQRCode(qrCode: String) {
        if (qrCodeDetected) return
        qrCodeDetected = true
        wireframe.navigate(QRCodeScanWireframeDestination.QRCode(qrCode))
    }
}